### ISSUE TITLE
Support only newer versions of Vim (8+) and NeoVim

### ISSUE THREAD

**Issue by Carpetsmoker** [fatih/vim-go#1847]
We currently [support Vim 7.4.1689 (March 2016) and newer](https://github.com/fatih/vim-go/blob/master/plugin/go.vim#L7). That number was chosen because that's what the most recent LTS version of Ubuntu shipped with at the time (16.04 LTS).

Ubuntu 18.04 LTS was released on 26th of April and ships with Vim 8.0.1453 (Feb 2018). We could consider using that version as the minimal version. Or perhaps Vim 8.0.0197 (Jan 2017) that ships with Ubuntu 17.10.

The upshot of this is that we can use newer features (such as `execute()`) as well as simplify some code paths.

Downside of this is that we *will* get issues/complains since many people clone master instead of use the last version. We even got some when we set the minimal version to 7.4.1689.

8.0.1453 is perhaps a little bit *too* new even for non-Ubuntu users, but 8.0.0197 seems pretty reasonable to me. It's already a year and a half old.

Thoughts @fatih @bhcleek?

**Comment by bhcleek** [comment]
I'm onboard!

**Comment by fatih** [comment]
I'm also ok because everyone is still able to use the plugin by setting `let g:go_version_warning = 0` and still able to work with it. Just be in mind that people still use old version of Vim and will continue doing it. We need to be sure we don't upset people. 

(offtopic: It's probably very hard, but It would be great to collect metrics from users, such as Go and Vim version. But I know this will be very hard to impossible due privacy concerns and other additional work on our side (collecting and storing data))

**Comment by bhcleek** [comment]
I agree that it would be nice to have some metrics, but I'm not quite sure how to get them in a way that's acceptable to users. 🤔 

If we were to drop support for vim74, then I'd expect to remove the synchronous code paths that are only used by vim74, but doing so would actually break vim-go on vim74. If you're not comfortable with that, then we should continue to officially support vim74 until you're comfortable removing those code paths.

**Comment by fatih** [comment]
My idea was not to remove imho. I know that a lot of people use vim-go from their remote servers, some of them don't have the option to update it. Also unlike Atom,Vscode, etc.. we don't have the option to auto-update Vim, so it'll be a long way until users switch to a newer version. However, of course, our focus should be on supporting Vim 8 and the newer API. If a problem comes up with Vim 7.4.x we should point and explain the reason why it's not our priority anymore. I think we're doing a great work already on this for the last years and people know we're treating everyone nice and good :)  

 > then I'd expect to remove the synchronous code paths that are only used by vim74, but doing so would actually break vim-go on vim74. 

Let's do this next year so people slowly adopt to this new mindset. I think once we remove all  Vim 7.4.x support, we can release vim-go as `v2.0.0` to show that it's breaking change. I don't have any dedicated plan on when and how we should do this, but just that we *could* do it.

**Comment by Carpetsmoker** [comment]
> everyone is still able to use the plugin by setting let g:go_version_warning = 0 and still able to work with it

Not if we start using newer stuff (such as `execute()`). *Some* stuff *may* work, but the more we use new stuff, the more things will start breaking.

> I know that a lot of people use vim-go from their remote servers, some of them don't have the option to update it.

It's fairly easy to get a newer Vim on a remote server, even without access to a compiler (just compile it on a different machine, upload binary+runtime files in `~/vim`). I used to do it all the time. Besides, no one is forcing anyone to use the latest version of vim-go.

> I think once we remove all Vim 7.4.x support, we can release vim-go as v2.0.0 to show that it's breaking change

The thing is that Vim doesn't really release software like that. Every patch applied is a new version (7.4.2000, 7.4.2001, etc.). At some point Bram said "This is now 8.0", but that could also have been 7.4.2367. There is no real hard distinction between "7.4" and "8.0". To use the jobs API for example you actually need 8.0.87, since versions before that don't work too well, so someone could have 86 different "8.0" versions and still need the sync code path.

---

Personally, I think that for vim-go to support Vim versions that are well over two years old is silly. Do you remember writing websites to be compatible with IE6? It kind of feels like that.

I'm not in favour of needlessly breaking stuff for people, but 7.4.2008 is from July 2016 – almost two years ago! I've wanted to use `execute()` on a number of occasions. It seems to me that if we're increasing the version to 7.4.2008 we might as well set it to 8.0.0197 (Jan 2017), as a lot of new stuff was added/fixed in those 6 months. We don't need to start dropping support for anything just yet, just having the "yo, you're going to need to update your Vim"-message out there seems useful to me.
